File: backend/app/services/session_namer.py
import re
from typing import Optional
from app.core import settings

AMBIGUOUS_PATTERNS = [
    re.compile(r"^(?:doc|document|scan|file|contract|agreement|test|sample|temp|upload|untitled|page|pdf|draft|new)[\d_\-\s.]*$", re.I),
    re.compile(r"^[0-9a-f]{8,}(?:-[0-9a-f]{4,})*$", re.I),
    re.compile(r"^\d{4,}[\d_\-\s]*$"),
    re.compile(r"^[a-z0-9]{1,3}$", re.I),
]

def is_ambiguous_or_unruly(stem: str) -> bool:
    clean = stem.strip()
    if len(clean) < 4:
        return True
    return any(p.match(clean) for p in AMBIGUOUS_PATTERNS)

def heuristic_session_title(filename: str, session_index: int = 1) -> str:
    if not filename:
        return f"Audit {session_index} – Contract" if session_index > 1 else "Contract Audit Session"

    stem = re.sub(r"\.[^/.]+$", "", filename).strip()

    if is_ambiguous_or_unruly(stem):
        clean_stem = re.sub(r"[_\-]+", " ", stem).strip() or "Contract"
        return f"Audit {session_index} – {clean_stem}" if session_index > 1 else f"Audit – {clean_stem}"

    # Clean noise tokens
    working = re.sub(r"\b(?:v\d+|version\s*\d+|final|draft|signed|executed|docusign|copy|scanned|revised|amended)\b", " ", stem, flags=re.I)
    working = re.sub(r"\b(?:20\d{2}|19\d{2})\b", " ", working)

    segments = [s.strip() for s in re.split(r"[_\-]+", working) if s.strip()]
    if not segments:
        return f"Audit {session_index} – {stem}" if session_index > 1 else f"Audit – {stem}"

    title = " ".join([s[0].upper() + s[1:] if len(s) > 1 else s.upper() for s in segments])
    if session_index > 1:
        return f"{title} (Audit {session_index})"
    return title

async def generate_smart_session_title(filename: str, session_index: int = 1) -> str:
    """
    AI Plugin for automatically and uniquely naming audit chat sessions based on filename.
    """
    stem = re.sub(r"\.[^/.]+$", "", filename or "").strip()
    if is_ambiguous_or_unruly(stem):
        return heuristic_session_title(filename, session_index)

    # Attempt fast LLM inference with fallback
    try:
        if settings.GROQ_API_KEY:
            from app.services.llm_service import GroqLLMService
            llm_service = GroqLLMService()
            llm = llm_service.get_llm(model=settings.FAST_GROQ_MODEL, temperature=0.1, max_tokens=35)
            prompt = (
                "You are an AI assistant that creates concise, professional audit titles for legal contracts. "
                "Given this file name, output a natural legal title (such as 'HarborPeak & Northstar Logistics Agreement' or 'Commercial Office Lease'). "
                "Do NOT include file extensions, version tags, or phrases like 'Audit -'. Return ONLY the title.\n"
                f"Filename: {filename}\nTitle:"
            )
            response = await llm.ainvoke(prompt)
            content = response.content.strip().strip('"').strip("'")
            if content and len(content) > 3 and "\n" not in content and len(content) < 60:
                if session_index > 1:
                    return f"{content} (Audit {session_index})"
                return content
    except Exception:
        pass

    return heuristic_session_title(filename, session_index)

File: frontend/src/utils/sessionNamer.ts
/**
 * AI Session Namer Plugin
 * 
 * Uniquely & automatically names legal audit/chat sessions based on filenames.
 * Handles company/party extraction and contract classification.
 * Falls back to "Audit [n] – <filename>" when filename is ambiguous, unruly, or generic.
 */

// Common generic or ambiguous prefixes/stems
const AMBIGUOUS_PATTERNS = [
  /^(?:doc|document|scan|file|contract|agreement|test|sample|temp|upload|untitled|page|pdf|draft|new)[\d_\-\s.]*$/i,
  /^[0-9a-f]{8,}(?:-[0-9a-f]{4,})*$/i, // UUID or Hex hashes
  /^\d{4,}[\d_\-\s]*$/,               // Mostly numbers (e.g. 20240923_001)
  /^[a-z0-9]{1,3}$/i,                 // Extremely short stems
];

// Common noise tokens to strip
const NOISE_TOKENS = [
  /\b(?:v\d+|version\s*\d+|final|draft|signed|executed|docusign|copy|scanned|stamped|revised|rev\d*|amended)\b/gi,
  /\b(?:20\d{2}|19\d{2})\b/g, // year numbers like 2024
];

// Known contract types to standardize
const CONTRACT_TYPES: Record<string, string> = {
  logistics: 'Logistics Services Agreement',
  'logistics services': 'Logistics Services Agreement',
  lease: 'Lease Agreement',
  'commercial lease': 'Commercial Lease Agreement',
  'commercial office lease': 'Commercial Office Lease',
  office_lease: 'Office Lease Agreement',
  nda: 'Non-Disclosure Agreement',
  'mutual nda': 'Mutual NDA',
  'non-disclosure': 'Non-Disclosure Agreement',
  'non disclosure': 'Non-Disclosure Agreement',
  employment: 'Employment Agreement',
  consulting: 'Consulting Services Agreement',
  msa: 'Master Services Agreement',
  'master services': 'Master Services Agreement',
  saas: 'SaaS Agreement',
  license: 'License Agreement',
  licensing: 'Licensing Agreement',
  supply: 'Supply Agreement',
  vendor: 'Vendor Agreement',
  settlement: 'Settlement Agreement',
  partnership: 'Partnership Agreement',
  indemnity: 'Indemnity Agreement',
  services: 'Services Agreement',
};

/**
 * Checks whether a filename is ambiguous, unruly, or lacks meaningful legal context.
 */
export function isAmbiguousOrUnruly(rawStem: string): boolean {
  const clean = rawStem.trim();
  if (clean.length < 4) return true;
  return AMBIGUOUS_PATTERNS.some((pat) => pat.test(clean));
}

/**
 * Generates an intelligent, unique, and natural session title based on a contract filename.
 * 
 * @param filename - The original contract file name (e.g. HarborPeak_Northstar_Logistics_Services_Agreement.pdf)
 * @param sessionIndex - Suffix index if multiple sessions exist on the same contract (1, 2, 3...)
 * @returns Clean, professional audit session title without artificial trimming
 */
export function generateSmartSessionTitle(filename: string, sessionIndex: number = 1): string {
  if (!filename) {
    return sessionIndex > 1 ? `Audit ${sessionIndex} – Contract` : 'Contract Audit Session';
  }

  // 1. Strip file extension
  const rawStem = filename.replace(/\.[^/.]+$/, '').trim();

  // 2. Check if the filename is ambiguous or unruly (e.g., scan001, doc_12, random hashes)
  if (isAmbiguousOrUnruly(rawStem)) {
    // Default fallback required by user: "Audit – <filename>" or "Audit n – <filename>"
    const cleanStem = rawStem.replace(/[_\-]+/g, ' ').trim() || 'Contract';
    return sessionIndex > 1 ? `Audit ${sessionIndex} – ${cleanStem}` : `Audit – ${cleanStem}`;
  }

  // 3. Clean noise tokens (e.g., _final, _signed, _v2)
  let working = rawStem;
  for (const pat of NOISE_TOKENS) {
    working = working.replace(pat, ' ');
  }

  // Split on delimiters (underscores, dashes)
  const segments = working
    .split(/[_\-]+/)
    .map((s) => s.trim())
    .filter(Boolean);

  if (segments.length === 0) {
    return sessionIndex > 1 ? `Audit ${sessionIndex} – ${rawStem}` : `Audit – ${rawStem}`;
  }

  // 4. Check for known contract types in segments
  const fullWorkingLower = segments.join(' ').toLowerCase();
  let matchedType = '';
  for (const [key, label] of Object.entries(CONTRACT_TYPES)) {
    if (fullWorkingLower.includes(key)) {
      matchedType = label;
      break;
    }
  }

  // 5. Extract parties (words before contract keywords or first 1-2 distinct words)
  const partyTokens: string[] = [];
  const contractKeywords = new Set([
    'agreement', 'contract', 'lease', 'nda', 'services', 'logistics',
    'commercial', 'employment', 'consulting', 'msa', 'license', 'supply',
    'settlement', 'partnership', 'terms', 'service'
  ]);

  for (const seg of segments) {
    const subWords = seg.replace(/([a-z])([A-Z])/g, '$1 $2').split(/\s+/);
    for (const w of subWords) {
      if (contractKeywords.has(w.toLowerCase())) {
        break;
      }
      if (w.length > 1 && !/^\d+$/.test(w)) {
        partyTokens.push(w);
      }
    }
  }

  let title = '';

  if (partyTokens.length >= 2 && matchedType) {
    const p1 = partyTokens[0];
    const p2 = partyTokens[1];
    title = `${p1} & ${p2} ${matchedType}`;
  } else if (partyTokens.length === 1 && matchedType) {
    title = `${partyTokens[0]} ${matchedType}`;
  } else if (matchedType) {
    const prefix = segments.slice(0, 2).join(' ');
    title = prefix.toLowerCase().includes(matchedType.toLowerCase().slice(0, 5))
      ? matchedType
      : `${prefix} ${matchedType}`;
  } else {
    title = segments
      .map((s) => s.charAt(0).toUpperCase() + s.slice(1))
      .join(' ');
  }

  title = title.replace(/\s+/g, ' ').replace(/& &/g, '&').trim();

  if (!title || title.length < 3) {
    title = rawStem.replace(/[_\-]+/g, ' ').trim();
  }

  // 6. Handle multiple sessions on the same contract uniquely
  if (sessionIndex > 1) {
    return `${title} (Audit ${sessionIndex})`;
  }

  return title;
}
